### PR TITLE
Multi tenancy api key support

### LINKED ISSUE #72: (User provided API keys) Take the OpenAI API load off the server host (make this a multi-tenant bot)
Currently, the project has one OpenAI token defined in an `.env` file, this token is the token that will be used for all requests made by anyone using the bot, across all the servers the bot is installed in. This is not viable, as this puts all the API burden on one person.

I'd like to shortly make this such that if a setting is toggled, every individual user is required to enter their own API key, and that key is what will be used for all requests in the service. The usage commands should also be updated to reflect per-user usage

### PR DESCRIPTION
**Pull request by Kav-K**
fixes #72

### CODE REVIEW

File: cogs/gpt_3_commands_and_converser.py [hunk 1]
```diff
@@ -1372,11 +1433,12 @@ async def settings(
...
 
 class ConversationView(discord.ui.View):
-    def __init__(self, ctx, converser_cog, id):
+    def __init__(self, ctx, converser_cog, id, custom_api_key=None):
         super().__init__(timeout=3600)  # 1 hour interval to redo.
         self.converser_cog = converser_cog
         self.ctx = ctx
-        self.add_item(RedoButton(self.converser_cog))
+        self.custom_api_key= custom_api_key
+        self.add_item(RedoButton(self.converser_cog, self.custom_api_key))
```
Comment (Hikari-Haru) [inline cogs/gpt_3_commands_and_converser.py:1441]: I'd want the api_key here keyworded when sent to redobutton

File: cogs/gpt_3_commands_and_converser.py [hunk 2]
```diff
@@ -932,12 +967,12 @@ async def encapsulated_send(
...
                         await ctx.respond(
                             response_text,
-                            view=ConversationView(ctx, self, ctx.channel.id),
+                            view=ConversationView(ctx, self, ctx.channel.id, custom_api_key),
                         )
                         if from_context
                         else await ctx.reply(
                             response_text,
-                            view=ConversationView(ctx, self, ctx.channel.id),
+                            view=ConversationView(ctx, self, ctx.channel.id, custom_api_key),
```
Comment (Hikari-Haru) [inline cogs/gpt_3_commands_and_converser.py:975]: These two views should be keyworded for sending the api

File: cogs/image_prompt_optimizer.py [hunk 3]
```diff
@@ -130,21 +142,23 @@ async def optimize(self, ctx: discord.ApplicationContext, prompt: str):
...
+    def __init__(self, converser_cog, image_service_cog, deletion_queue, custom_api_key=None):
         super().__init__(timeout=None)
         self.cog = converser_cog
         self.image_service_cog = image_service_cog
         self.deletion_queue = deletion_queue
-        self.add_item(RedoButton(self.cog, self.image_service_cog, self.deletion_queue))
-        self.add_item(DrawButton(self.cog, self.image_service_cog, self.deletion_queue))
+        self.custom_api_key = custom_api_key
+        self.add_item(RedoButton(self.cog, self.image_service_cog, self.deletion_queue, self.custom_api_key))
+        self.add_item(DrawButton(self.cog, self.image_service_cog, self.deletion_queue, self.custom_api_key))
```
Comment (Hikari-Haru) [inline cogs/image_prompt_optimizer.py:152]: Should be keyworded here too

**Comment by Kav-K** [comment]
Am looking for some more reviews on the reliability and stability of this before I merge, I've done some local testing and things look good but I'm looking for more input :)

**Comment by Kav-K** [comment]
fixes #72

**Review by Joepetey — COMMENTED** [review]
Looks good to me!

**Pull request merged**